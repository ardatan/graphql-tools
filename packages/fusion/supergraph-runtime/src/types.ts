import { DocumentNode } from 'graphql';

// Query planner types
export type PlanNode = ResolveNode | SequenceNode | ParallelNode;

export interface ResolveNode {
  type: 'Resolve';
  subgraph: string;
  document: DocumentNode;
  provided?: {
    selections?: Map<string, string[]>;
    variablesInSelections?: Map<string, Map<string, string[]>>;
    selectionFields?: Map<string, Map<string, string[]>>;
  };
  required?: {
    variables?: string[];
    selections?: Map<string, string[]>;
  };
  batch?: boolean;
}

export interface SequenceNode {
  type: 'Sequence';
  nodes: PlanNode[];
}

export interface ParallelNode {
  type: 'Parallel';
  nodes: PlanNode[];
}

// Schema types
export type ResolverKind = 'FETCH' | 'BATCH' | 'SUBSCRIBE';

export interface ResolverVariableConfig {
  name: string;
  select?: string;
  subgraph: string;
}

export interface ResolverConfig {
  operation: string;
  kind: ResolverKind;
  subgraph: string;
}
