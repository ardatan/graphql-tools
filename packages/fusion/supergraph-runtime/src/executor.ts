import type { GraphQLError } from 'graphql';
import _ from 'lodash';
import {
  inspect,
  isAsyncIterable,
  isPromise,
  type ExecutionRequest,
  type ExecutionResult,
  type Executor,
  type MaybePromise,
} from '@graphql-tools/utils';
import type { PlanNode } from './types.js';

export type QueryPlanExecutorRequest = {
  queryPlan: PlanNode;
} & Omit<ExecutionRequest, 'document'>;

export interface CreateQueryPlanExecutorOpts {
  subgraphExecutors: Map<string, Executor>;
}
export type QueryPlanExecutor = (
  request: QueryPlanExecutorRequest,
) => MaybePromise<ExecutionResult>;

function arrayGet(obj: any, path: string[], index?: number): any {
  const currentPath = path[0];
  if (path.length === 1) {
    if (Array.isArray(obj)) {
      if (index != null) {
        return obj[index][currentPath];
      }
      return obj.map(item => item[currentPath]);
    }
    return obj[currentPath];
  }
  if (Array.isArray(obj)) {
    if (index != null) {
      return arrayGet(obj[index][currentPath], path.slice(1));
    }
    return obj.map(item => arrayGet(item[currentPath], path.slice(1)));
  }
  return arrayGet(obj[currentPath], path.slice(1), index);
}

export function createQueryPlanExecutor({
  subgraphExecutors,
}: CreateQueryPlanExecutorOpts): QueryPlanExecutor {
  return function executeQueryPlan({ queryPlan, variables, context }) {
    const variablesState = new Map<string, any>(Object.entries(variables ?? {}));
    const multiVariablesState = new Map<string, any[]>();
    const selectionsState = new Map<string, any>();
    const errors: GraphQLError[] = [];
    let finalResult: any = null;
    function _executeNode(node: PlanNode, index?: number): any {
      switch (node.type) {
        case 'Resolve': {
          let executorVariables: Record<string, any> | undefined;
          if (node.required?.variables) {
            for (const variableName of node.required.variables) {
              let variableValue = variablesState.get(variableName);
              if (variableValue == null) {
                const multiVariableValue = multiVariablesState.get(variableName);
                if (multiVariableValue != null) {
                  if (node.batch) {
                    variableValue = multiVariableValue;
                  } else {
                    if (index == null) {
                      return Promise.all(multiVariableValue.map((_, i) => _executeNode(node, i)));
                    }
                    variableValue = multiVariableValue[index];
                  }
                }
              }
              executorVariables ||= {};
              executorVariables[variableName] = variableValue;
            }
          }
          const subgraphExecutor = subgraphExecutors.get(node.subgraph);
          if (!subgraphExecutor) {
            throw new Error(`No executor found for subgraph ${node.subgraph}`);
          }
          const result$ = subgraphExecutor({
            document: node.document,
            context,
            variables: executorVariables,
          });
          const handleResult = (result: ExecutionResult) => {
            if (result.errors) {
              errors.push(...result.errors);
            }
            if (node.required?.selections) {
              for (const [selectionName, selectionPathInResult] of node.required.selections) {
                const selectionToBeSet = arrayGet(result.data, selectionPathInResult, index);
                if (selectionToBeSet == null) {
                  throw new Error(
                    `Missing selection ${selectionPathInResult} in ${inspect(result.data)}`,
                  );
                }
                selectionsState.set(selectionName, selectionToBeSet);
                if (node.provided?.variablesInSelections) {
                  const isArraySelection = Array.isArray(selectionToBeSet);
                  const variablesInSelection =
                    node.provided.variablesInSelections.get(selectionName);
                  if (variablesInSelection) {
                    for (const [variableName, variablePathInSelection] of variablesInSelection) {
                      if (isArraySelection) {
                        let multiVariable = multiVariablesState.get(variableName);
                        if (!multiVariable) {
                          multiVariable = [];
                          multiVariablesState.set(variableName, multiVariable);
                        }
                        for (const selectionItem of selectionToBeSet) {
                          const variableToBeSet = _.get(selectionItem, variablePathInSelection);
                          multiVariable.push(variableToBeSet);
                        }
                      } else {
                        const variableToBeSet = _.get(selectionToBeSet, variablePathInSelection);
                        variablesState.set(variableName, variableToBeSet);
                      }
                    }
                  }
                }
              }
            }
            if (node.provided?.selections) {
              for (const [selectionName, selectionPathInResult] of node.provided.selections) {
                const selection = selectionsState.get(selectionName);
                if (!selection) {
                  throw new Error(`Missing selection ${selectionName}`);
                }
                const selectionToBeSet = _.get(result.data, selectionPathInResult);
                if (!selectionToBeSet) {
                  throw new Error(`Missing selection exported ${selectionPathInResult}`);
                }
                if (node.batch) {
                  const isArraySelection = Array.isArray(selection);
                  for (const i in selectionToBeSet) {
                    Object.assign(isArraySelection ? selection[i] : selection, selectionToBeSet[i]);
                  }
                } else if (index != null) {
                  Object.assign(selection[index], selectionToBeSet);
                } else {
                  Object.assign(selection, selectionToBeSet);
                }
              }
            }
            if (node.provided?.selectionFields) {
              for (const [selectionName, selectionFieldMap] of node.provided.selectionFields) {
                const selection = selectionsState.get(selectionName);
                if (!selection) {
                  throw new Error(`Missing selection ${selectionName}`);
                }
                for (const [pathInSelection, pathInResult] of selectionFieldMap) {
                  const fieldToBeSet = _.get(result.data, pathInResult);
                  if (node.batch) {
                    for (const i in fieldToBeSet) {
                      _.set(selection[i], pathInSelection, fieldToBeSet[i]);
                    }
                  } else if (index != null) {
                    _.set(selection[index], pathInSelection, fieldToBeSet);
                  } else {
                    _.set(selection, pathInSelection, fieldToBeSet);
                  }
                }
              }
            }
            if (!finalResult) {
              finalResult = result.data;
            }
          };
          if (isAsyncIterable(result$)) {
            return Promise.resolve().then(async () => {
              for await (const result of result$) {
                handleResult(result);
              }
            });
          }
          if (isPromise(result$)) {
            return result$.then(maybeAsyncIterableResult => {
              if (isAsyncIterable(maybeAsyncIterableResult)) {
                return Promise.resolve().then(async () => {
                  for await (const result of maybeAsyncIterableResult) {
                    handleResult(result);
                  }
                });
              }
              handleResult(maybeAsyncIterableResult);
            });
          }
          handleResult(result$);
          break;
        }
        case 'Parallel': {
          const promises: PromiseLike<any>[] = [];
          for (const nodeElem of node.nodes) {
            const res$ = _executeNode(nodeElem, index);
            if (isPromise(res$)) {
              promises.push(res$);
            }
          }
          if (promises.length) {
            return Promise.all(promises);
          }
          break;
        }
        case 'Sequence': {
          const nodeIterator = node.nodes[Symbol.iterator]();
          // eslint-disable-next-line no-inner-declarations
          function executeNextNode(): any {
            const nextNode = nodeIterator.next();
            if (nextNode.done) {
              return;
            }
            const res$ = _executeNode(nextNode.value);
            if (isPromise(res$)) {
              return res$.then(executeNextNode);
            }
            return executeNextNode();
          }
          return executeNextNode();
        }
      }
    }
    _executeNode(queryPlan);
    return {
      errors,
      data: finalResult,
    };
  };
}
