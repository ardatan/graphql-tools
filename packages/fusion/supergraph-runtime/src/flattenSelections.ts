import { FieldNode, FragmentDefinitionNode, Kind, SelectionNode, SelectionSetNode } from 'graphql';

export interface FlattenedFieldNode extends FieldNode {
  selectionSet?: FlattenedSelectionSet;
}

export interface FlattenedSelectionSet extends SelectionSetNode {
  selections: FlattenedFieldNode[];
}

export function flattenSelections(
  selections: readonly SelectionNode[],
  fragments: Record<string, FragmentDefinitionNode>,
): FlattenedFieldNode[] {
  return selections.flatMap(selection => {
    switch (selection.kind) {
      case Kind.FIELD:
        return [
          {
            ...selection,
            selectionSet: selection.selectionSet && {
              ...selection.selectionSet,
              selections: flattenSelections(selection.selectionSet.selections, fragments),
            },
          },
        ];
      case Kind.INLINE_FRAGMENT:
        return flattenSelections(selection.selectionSet.selections, fragments);
      case Kind.FRAGMENT_SPREAD: {
        const fragment = fragments[selection.name.value];
        if (!fragment) {
          throw new Error(`No fragment found for ${selection.name.value}`);
        }
        return flattenSelections(fragment.selectionSet.selections, fragments);
      }
      default:
        throw new Error(`Unexpected selection node kind ${(selection as any).kind}`);
    }
  });
}
