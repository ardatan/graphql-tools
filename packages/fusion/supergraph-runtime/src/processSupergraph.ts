import { ASTVisitFn, buildASTSchema, GraphQLSchema, Kind, parse, visit } from 'graphql';
import { NamedDefinitionNodeWithDirectives } from './types.js';

export interface ProcessedSupergraph {
  supergraphSchema: GraphQLSchema;
  subgraphSchemas: Map<string, GraphQLSchema>;
}

export function processSupergraphSdl(supergraphSdl: string): ProcessedSupergraph {
  const subgraphNames = new Set<string>();
  const supergraphAst = parse(supergraphSdl, { noLocation: true });
  visit(supergraphAst, {
    [Kind.DIRECTIVE](node) {
      const subgraphArgument = node.arguments?.find(arg => arg.name.value === 'subgraph');
      if (subgraphArgument && subgraphArgument.value.kind === Kind.STRING) {
        subgraphNames.add(subgraphArgument.value.value);
      }
    },
  });

  const subgraphSchemas = new Map<string, GraphQLSchema>();

  // for (const subgraphName of subgraphNames) {
  //   const subgraphSchemaAst = visit(
  //     supergraphAst,
  //     Object.fromEntries(
  //       sourceDirectiveLocations.map(directiveLocation => [
  //         directiveLocation,
  //         {
  //           enter: createFilterBySourceDirectiveVisitorSubgraphName(subgraphName),
  //         },
  //       ]),
  //     ),
  //   );
  //   const subgraphSchema = buildASTSchema(subgraphSchemaAst);
  //   subgraphSchemas.set(subgraphName, subgraphSchema);
  // }
  const supergraphSchema = buildASTSchema(supergraphAst);

  return {
    supergraphSchema,
    subgraphSchemas,
  };
}

const sourceDirectiveLocations = [
  'ObjectTypeDefinition',
  'FieldDefinition',
  'EnumTypeDefinition',
  'EnumValueDefinition',
  'InputObjectTypeDefinition',
  'InputFieldDefinition',
  'ScalarTypeDefinition',
] as const;

function createFilterBySourceDirectiveVisitorSubgraphName(
  subgraphName: string,
): ASTVisitFn<NamedDefinitionNodeWithDirectives> {
  return function filterBySourceDirectiveVisitor(node) {
    if (
      node.directives?.some(
        directiveNode =>
          directiveNode.arguments?.some(
            argumentNode =>
              argumentNode.name.value === 'subgraph' &&
              argumentNode.value.kind === Kind.STRING &&
              argumentNode.value.value === subgraphName,
          ),
      )
    ) {
      return node;
    }
    return null;
  };
}
