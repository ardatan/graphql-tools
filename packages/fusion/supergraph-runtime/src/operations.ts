import {
  DocumentNode,
  FragmentDefinitionNode,
  GraphQLSchema,
  Kind,
  OperationDefinitionNode,
} from 'graphql';
import { Executor, getRootTypeMap } from '@graphql-tools/utils';
import {
  createExecutableResolverOperationNodesWithDependencyMap,
  executeResolverOperationNodesWithDependenciesInParallel,
} from './execution.js';
import { FlattenedFieldNode, flattenSelections } from './flattenSelections.js';
import { visitFieldNodeForTypeResolvers } from './query-planning.js';

export function planOperation(
  supergraph: GraphQLSchema,
  document: DocumentNode,
  operationName?: string,
) {
  let operationAst: OperationDefinitionNode | undefined;
  const fragments: Record<string, FragmentDefinitionNode> = Object.create(null);
  for (const definition of document.definitions) {
    if (definition.kind === 'OperationDefinition') {
      if (!operationName && operationAst) {
        throw new Error('Must provide operation name if query contains multiple operations');
      }
      if (!operationName || definition.name?.value === operationName) {
        operationAst = definition;
      }
    } else if (definition.kind === 'FragmentDefinition') {
      fragments[definition.name.value] = definition;
    }
  }

  if (!operationAst) {
    throw new Error(`No operation found with name ${operationName}`);
  }

  const flattenedFakeFieldNode: FlattenedFieldNode = {
    kind: Kind.FIELD,
    name: {
      kind: Kind.NAME,
      value: '__fake',
    },
    selectionSet: {
      kind: Kind.SELECTION_SET,
      selections: flattenSelections(operationAst.selectionSet.selections, fragments),
    },
  };

  const rootTypeMap = getRootTypeMap(supergraph);
  const operationType = operationAst.operation;
  const rootType = rootTypeMap.get(operationType);
  if (!rootType) {
    throw new Error(`No root type found for operation type ${operationType}`);
  }

  return visitFieldNodeForTypeResolvers('ROOT', flattenedFakeFieldNode, rootType, supergraph, {
    currentVariableIndex: 0,
  });
}

export function executeOperation(
  supergraph: GraphQLSchema,
  executorMap: Map<string, Executor>,
  document: DocumentNode,
  operationName?: string,
) {
  const plan = planOperation(supergraph, document, operationName);
  const executablePlan = createExecutableResolverOperationNodesWithDependencyMap(
    plan.resolverOperationNodes,
    plan.resolverDependencyFieldMap,
  );

  return executeResolverOperationNodesWithDependenciesInParallel(
    executablePlan.newResolverOperationNodes,
    executablePlan.newResolverDependencyMap,
    new Map(),
    executorMap,
  );
}
