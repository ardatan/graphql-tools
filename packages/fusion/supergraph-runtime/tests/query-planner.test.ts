import { createQueryPlannerFromProcessedSupergraph } from "../src/queryPlanner";
import { parse, print } from "graphql";
import { PlanNode } from "../src/types";
import { inspect } from "util";
import { processSupergraphSdl } from "../src/processSupergraph";
import { createQueryPlanExecutor } from "../src/executor";
import { makeExecutableSchema } from "@graphql-tools/schema";
import { createDefaultExecutor } from "@graphql-tools/delegate";

// Follow test steps in Chillicream

describe('createQueryPlannerFromSupergraph', () => {
  it('test', async () => {
    const supergraphSdl = /* GraphQL */`
      directive @resolver(
        operation: String!
        subgraph: String!
      ) repeatable on OBJECT | FIELD_DEFINITION
      directive @source(
        subgraph: String!
      ) repeatable on OBJECT | FIELD_DEFINITION | ARGUMENT_DEFINITION
      directive @variable(
        name: String!
        select: String
        subgraph: String!
      ) repeatable on OBJECT | FIELD_DEFINITION
      type Query
          @source(subgraph: "A")
          @source(subgraph: "B") {
        foo(id: ID! @source(subgraph: "A") @source(subgraph: "B")): Foo!
          @source(subgraph: "A")
          @resolver(operation: "query RootFooA($Foo_id:ID!) { foo(id: $Foo_id) }", subgraph: "A")
          @source(subgraph: "B")
          @resolver(operation: "query RootFooB($Foo_id:ID!) { foo(id: $Foo_id) }", subgraph: "B")

        someOtherField(fooId: ID! @source(subgraph: "C")): String! @source(subgraph: "C")
      }

      type Foo
          @source(subgraph: "A")
          @source(subgraph: "B")
          @variable(name: "Foo_id", select: "id", subgraph: "A")
          @variable(name: "Foo_id", select: "id", subgraph: "B")
          @resolver(operation: "query ResolveFooFromA($Foo_id: ID!) { foo(id: $Foo_id) }", subgraph: "A")
          @resolver(operation: "query ResolveFooFromB($Foo_id: ID!) { foo(id: $Foo_id) }", subgraph: "B") {
        id: ID!
          @source(subgraph: "A")
          @source(subgraph: "B")

        bar: String!
          @source(subgraph: "A")
        baz: String!
          @source(subgraph: "B")

        qux: String!
          @source(subgraph: "A")

        child: Foo
          @source(subgraph: "B")

        otherField: String!
          @variable(name: "fooId", select: "id", subgraph: "A")
          @variable(name: "fooId", select: "id", subgraph: "B")
          @resolver(operation: "query RootSomeOtherFieldC($fooId:ID!) { someOtherField(fooId: $fooId) }", subgraph: "C")
      }
    `;
    const processedSupergraph = processSupergraphSdl(supergraphSdl);
    const queryPlanner = createQueryPlannerFromProcessedSupergraph(processedSupergraph);
    const plan = queryPlanner.plan({
      document: parse(/* GraphQL */`
        query Test($id: ID!) {
          foo(id: $id) {
            bar
            baz
            child {
              qux
              otherField
            }
          }
        }
      `),
      operationName: 'test',
    });
    console.log(inspect(serializePlanNode(plan), false, Infinity));
    const aSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */`
        type Query {
          foo(id: ID!): Foo!
        }

        type Foo {
          id: ID!
          bar: String!
          qux: String!
        }
      `,
      resolvers: {
        Query: {
          foo: () => ({
            id: 'A',
            bar: 'A',
            qux: 'A',
          }),
        },
      },
    })
    const bSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */`
        type Query {
          foo(id: ID!): Foo!
        }

        type Foo {
          id: ID!
          baz: String!
          child: Foo
        }
      `,
      resolvers: {
        Query: {
          foo: () => ({
            id: 'B',
            baz: 'B',
            child: {
              id: 'B',
              qux: 'B',
            },
          }),
        },
      },
    })
    const cSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */`
        type Query {
          someOtherField(fooId: ID!): String!
        }
      `,
      resolvers: {
        Query: {
          someOtherField: (_, { fooId }) => `C ${fooId}`,
        },
      },
    })
    const executor = createQueryPlanExecutor({
      subgraphExecutors: new Map([
        ['A', createDefaultExecutor(aSchema)],
        ['B', createDefaultExecutor(bSchema)],
        ['C', createDefaultExecutor(cSchema)],
      ]),
    });
    const result = await executor({
      queryPlan: plan,
      variables: {
        Foo_id: 'A',
      },
    });
    console.log(inspect(result, false, Infinity));
  })
})

function serializeMap<K extends string, V>(map: Map<K, V>): Record<K,V> {
  return Object.fromEntries(map.entries()) as Record<K,V>;
}

function serializePlanNode(planNode: PlanNode) {
  if (planNode.type === 'Resolve') {
    const serializedResolveNode: any = {
      type: 'Resolve',
      subgraph: planNode.subgraph,
      document: print(planNode.document),
    }
    if (planNode.provided) {
      serializedResolveNode.provided = {};
      if (planNode.provided.variables) {
        serializedResolveNode.provided.variables = serializeMap(planNode.provided.variables);
      }
      if (planNode.provided.selections) {
        serializedResolveNode.provided.selections = serializeMap(planNode.provided.selections);
      }
      if (planNode.provided.selectionFields) {
        serializedResolveNode.provided.selectionFields = planNode.provided.selectionFields;
      }
    }
    if (planNode.required) {
      serializedResolveNode.required = {};
      if (planNode.required.variables) {
        serializedResolveNode.required.variables = planNode.required.variables;
      }
      if (planNode.required.selections) {
        serializedResolveNode.required.selections = serializeMap(planNode.required.selections);
      }
    }
    return serializedResolveNode;
  }
  return {
    type: planNode.type,
    nodes: planNode.nodes.map(serializePlanNode),
  }
}
