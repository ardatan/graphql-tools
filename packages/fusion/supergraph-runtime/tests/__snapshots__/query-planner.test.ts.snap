// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Planner & Executor Lists creates a serializable query plan: query-result-Lists 1`] = `
{
  "nodes": [
    {
      "document": "query ResolveFoosFromB {
  foos {
    __selection__0__typename: __typename
    __variable__0: id
    baz
  }
}",
      "provided": {
        "variablesInSelections": Map {
          "__selection__0" => Map {
            "__variable__0" => [
              "__variable__0",
            ],
          },
        },
      },
      "required": {
        "selections": {
          "__selection__0": [
            "foos",
          ],
        },
      },
      "subgraph": "B",
      "type": "Resolve",
    },
    {
      "document": "query ResolveFooFromA($__variable__0: ID!) {
  __selection__0: foo(id: $__variable__0) {
    bar
  }
}",
      "provided": {
        "selections": {
          "__selection__0": [
            "__selection__0",
          ],
        },
      },
      "required": {
        "variables": [
          "__variable__0",
        ],
      },
      "subgraph": "A",
      "type": "Resolve",
    },
  ],
  "type": "Sequence",
}
`;

exports[`Query Planner & Executor Regular creates a serializable query plan: query-result-Regular 1`] = `
{
  "nodes": [
    {
      "document": "query RootFooA($my_foo_id: ID!) {
  foo(id: $my_foo_id) {
    __selection__0__typename: __typename
    __variable__0: id
    bar
  }
}",
      "provided": {
        "variablesInSelections": Map {
          "__selection__0" => Map {
            "__variable__0" => [
              "__variable__0",
            ],
          },
        },
      },
      "required": {
        "selections": {
          "__selection__0": [
            "foo",
          ],
        },
        "variables": [
          "my_foo_id",
        ],
      },
      "subgraph": "A",
      "type": "Resolve",
    },
    {
      "batch": true,
      "document": "query ResolveFoosFromB($__variable__0: [ID!]!) {
  __selection__0: foos(ids: $__variable__0) {
    baz
    child {
      __selection__1__typename: __typename
      __variable__2: id
      __selection__1__typename: __typename
      __variable__1: id
    }
  }
}",
      "provided": {
        "selections": {
          "__selection__0": [
            "__selection__0",
          ],
        },
        "variablesInSelections": Map {
          "__selection__1" => Map {
            "__variable__2" => [
              "__variable__2",
            ],
            "__variable__1" => [
              "__variable__1",
            ],
          },
        },
      },
      "required": {
        "selections": {
          "__selection__1": [
            "__selection__0",
            "child",
          ],
        },
        "variables": [
          "__variable__0",
        ],
      },
      "subgraph": "B",
      "type": "Resolve",
    },
    {
      "nodes": [
        {
          "document": "query ResolveFooFromA($__variable__1: ID!) {
  __selection__1: foo(id: $__variable__1) {
    qux
  }
}",
          "provided": {
            "selections": {
              "__selection__1": [
                "__selection__1",
              ],
            },
          },
          "required": {
            "variables": [
              "__variable__1",
            ],
          },
          "subgraph": "A",
          "type": "Resolve",
        },
        {
          "document": "query RootSomeOtherFieldC($__variable__2: ID!) {
  __field__otherField__selection__1: someOtherField(fooId: $__variable__2)
}",
          "provided": {
            "selectionFields": Map {
              "__selection__1" => Map {
                "otherField" => [
                  "__field__otherField__selection__1",
                ],
              },
            },
          },
          "required": {
            "variables": [
              "__variable__2",
            ],
          },
          "subgraph": "C",
          "type": "Resolve",
        },
      ],
      "type": "Parallel",
    },
  ],
  "type": "Sequence",
}
`;
